62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+
+#include "Hatch.h"
+using namespace nbl::hlsl;
+
+#include "common.hlsl"
+#include <nbl/builtin/hlsl/equations/cubic.hlsl>
+#include <nbl/builtin/hlsl/equations/quartic.hlsl>
+
+namespace hatchutils {
+    
+	// Intended to mitigate issues with NaNs and precision by falling back to using simpler functions when the higher roots are small enough
+	static std::array<double, 4> solveQuarticRoots(double a, double b, double c, double d, double e, double t_start, double t_end)
+	{
+		std::array<double, 4> t = { -1.0, -1.0, -1.0, -1.0 }; // only two candidates in range, ever
+
+		const double quadCoeffMag = std::max(std::abs(d), std::abs(e));
+		const double cubCoeffMag = std::max(std::abs(c), quadCoeffMag);
+		if (std::abs(a) > std::max(std::abs(b), cubCoeffMag) * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Quartic<double>::construct(a, b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 4);
+		}
+		else if (abs(b) > quadCoeffMag)
```

Comment:
You're missing a `* QUARTIC_THRESHOLD`
also make that var a static constexpr and define in locally in this function